{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multiomics BMI Paper — Visualization of Network Interaction Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***by Kengo Watanabe***  \n",
    "\n",
    "This Jupyter Notebook (with R kernel) visualized the analyte–analyte pairs that were significantly modified by MetBMI (i.e., GLM results) while highlighting the anayte–analyte pairs that were significantly modified by days in program (i.e., GEE results).  \n",
    "\n",
    "Input files:  \n",
    "* GEE result summary: 220806_Multiomics-BMI-NatMed1stRevision_LongitudinalNetworkAnalysis-ver2_GEE-interaction.tsv  \n",
    "* Arivale baseline blood omics: 210104_Biological-BMI-paper_data-cleaning-BMI-omics_baseline-\\[metDF/protDF/chemDF\\]-without-imputation_final-cohort.tsv  \n",
    "\n",
    "Output figures and tables:  \n",
    "* Figure 6a  \n",
    "\n",
    "Original notebook (memo for my future tracing):  \n",
    "* dalek:\\[JupyterLab HOME\\]/220621_Multiomics-BMI-NatMedRevision/220806_Multiomics-BMI-NatMed1stRevision_NetworkAnalysisVisualization.ipynb  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.6     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.7     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.9\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.2.0     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.1.2     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.15\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"circlize: 0.4.15\"\n",
      "[1] \"RColorBrewer: 1.1.3\"\n",
      "[1] \"colorspace: 2.0.3\"\n",
      "[1] \"arivale.data.interface: 0.0.0.1\"\n"
     ]
    }
   ],
   "source": [
    "library(\"tidyverse\")\n",
    "options(repr.plot.width=5, repr.plot.height=5)#Default=7x7\n",
    "\n",
    "#CRAN\n",
    "for (package in c(\"circlize\", \"RColorBrewer\", \"colorspace\")) {\n",
    "    #install.packages(package)\n",
    "    eval(bquote(library(.(package))))\n",
    "    print(str_c(package, \": \", as.character(packageVersion(package))))\n",
    "}\n",
    "\n",
    "#Internal\n",
    "for (package in c(\"arivale.data.interface\")) {\n",
    "    eval(bquote(library(.(package))))\n",
    "    print(str_c(package, \": \", as.character(packageVersion(package))))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Longitudinal network analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Use all-omics version; but majority is metabolomics.  \n",
    "> –> Use subcategory for metabolomics.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-1. Prepare the significant pairs in GLM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import significant analyte pairs in GLM with the subsequent GEE results\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"220806_Multiomics-BMI-NatMed1stRevision_LongitudinalNetworkAnalysis-ver2_\"\n",
    "fileName <- \"GEE-interaction.tsv\"\n",
    "pair.tbl <- read_delim(str_c(fileDir, ipynbName, fileName), delim=\"\\t\") %>%\n",
    "    dplyr::select(PairLabel, Variable1, Variable2, Pearson_r,\n",
    "                  GLM_Bcoef, GLM_AdjPval, GEE_Bcoef, GEE_AdjPval) %>%\n",
    "    dplyr::filter(GLM_AdjPval < 0.05)\n",
    "temp <- unique(c(pair.tbl$Variable1, pair.tbl$Variable2))\n",
    "print(str_c(\"Significant pairwise alnalyte interactions with MetBMI (FDR < 0.05): \", nrow(pair.tbl)))\n",
    "print(str_c(\" -> nAnalytes: \", length(temp)))\n",
    "\n",
    "temp.tbl <- pair.tbl %>% dplyr::filter(GEE_AdjPval < 0.05)\n",
    "temp <- unique(c(temp.tbl$Variable1, temp.tbl$Variable2))\n",
    "print(str_c(\"Significant pairwise analyte interactions with days (FDR < 0.05): \", nrow(temp.tbl)))\n",
    "print(str_c(\" -> nAnalytes: \", length(temp)))\n",
    "pair.tbl <- pair.tbl %>% dplyr::mutate(Highlight=ifelse((GEE_AdjPval < 0.05), 1, 0))\n",
    "\n",
    "#Convert to -log10 for potential edge attribution\n",
    "pair.tbl <- pair.tbl %>% dplyr::mutate(GLM_AdjPval_scaled=-log10(GLM_AdjPval))\n",
    "\n",
    "head(pair.tbl)\n",
    "tail(pair.tbl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-2. Prepare the analyte in the significant paris"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare node table with omics category\n",
    "node.tbl <- tibble(Analyte=unique(c(pair.tbl$Variable1, pair.tbl$Variable2))) %>%\n",
    "    dplyr::mutate(Category=rep(0, n()))#Dummy\n",
    "    \n",
    "temp <- c(\"public_client_id\", \"log_BaseBMI\", \"Sex\", \"BaseAge\", \"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\", \"Race\")\n",
    "fileDir <- \"../210104_Biological-BMI-paper/ExportData/\"\n",
    "ipynbName <- \"210104_Biological-BMI-paper_data-cleaning-BMI-omics_\"\n",
    "fileName <- \"baseline-metDF-without-imputation_final-cohort.tsv\"\n",
    "temp.c <- read_delim(str_c(fileDir, ipynbName, fileName), delim=\"\\t\") %>%\n",
    "    dplyr::select(-temp) %>%\n",
    "    names(.)\n",
    "print(str_c(\"Metabolomics: \", length(temp.c)))\n",
    "node.tbl <- node.tbl %>% dplyr::mutate(Category=ifelse((Analyte %in% temp.c), \"Metabolomics\", Category))\n",
    "\n",
    "fileName <- \"baseline-protDF-without-imputation_final-cohort.tsv\"\n",
    "temp.c <- read_delim(str_c(fileDir, ipynbName, fileName), delim=\"\\t\") %>%\n",
    "    dplyr::select(-temp) %>%\n",
    "    names(.)\n",
    "print(str_c(\"Proteomics: \", length(temp.c)))\n",
    "node.tbl <- node.tbl %>% dplyr::mutate(Category=ifelse((Analyte %in% temp.c), \"Proteomics\", Category))\n",
    "\n",
    "fileName <- \"baseline-chemDF-without-imputation_final-cohort.tsv\"\n",
    "temp.c <- read_delim(str_c(fileDir, ipynbName, fileName), delim=\"\\t\") %>%\n",
    "    dplyr::select(-temp) %>%\n",
    "    names(.)\n",
    "print(str_c(\"Clinical labs: \", length(temp.c)))\n",
    "node.tbl <- node.tbl %>% dplyr::mutate(Category=ifelse((Analyte %in% temp.c), \"Clinical labs\", Category))\n",
    "\n",
    "#Check\n",
    "node.tbl %>% dplyr::filter(Category==0)\n",
    "print(str_c(\"nrow: \", nrow(node.tbl)))\n",
    "head(node.tbl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Significant analyte in GEE (FDR < 0.05)\n",
    "temp.tbl <- pair.tbl %>% dplyr::filter(GEE_AdjPval < 0.05)\n",
    "temp <- unique(c(temp.tbl$Variable1, temp.tbl$Variable2))\n",
    "node.tbl <- node.tbl %>% dplyr::mutate(Highlight=ifelse((Analyte %in% temp), 1, 0))\n",
    "\n",
    "#The number of connected edges\n",
    "temp <- c(pair.tbl$Variable1, pair.tbl$Variable2)\n",
    "node.tbl <- node.tbl %>%\n",
    "    dplyr::mutate(nEdges=sapply(Analyte, function(x){length(temp[which(temp==x)])})) %>%\n",
    "    dplyr::arrange(desc(nEdges))\n",
    "\n",
    "head(node.tbl)\n",
    "\n",
    "#Summary\n",
    "temp.tbl <- node.tbl %>%\n",
    "    dplyr::group_by(Category) %>%\n",
    "    dplyr::summarize(Count=n()) %>%\n",
    "    dplyr::arrange(desc(Count))\n",
    "temp.tbl\n",
    "temp.tbl <- node.tbl %>%\n",
    "    dplyr::filter(Highlight==1) %>%\n",
    "    dplyr::group_by(Category) %>%\n",
    "    dplyr::summarize(Count=n()) %>%\n",
    "    dplyr::arrange(desc(Count))\n",
    "temp.tbl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-3. Prepare the metabolomics pathway information for subcategory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the Arivale metabolomics metadata for the significant analytes in GLM\n",
    "snapshotPath <- \"/shared-data/snapshots/arivale_snapshot_ISB_2019-05-19_1330\"\n",
    "node.tbl <- get_snapshot(\"metabolomics_metadata\", path=snapshotPath) %>%\n",
    "    as_tibble() %>%\n",
    "    dplyr::select(BIOCHEMICAL_NAME, SUPER_PATHWAY, SUB_PATHWAY) %>%\n",
    "    dplyr::rename(Analyte=BIOCHEMICAL_NAME) %>%\n",
    "    dplyr::left_join(node.tbl, ., by=\"Analyte\")\n",
    "\n",
    "#Fill empty as string \"NA\"\n",
    "node.tbl <- node.tbl %>%\n",
    "    dplyr::mutate(SUPER_PATHWAY=ifelse(SUPER_PATHWAY==\"\", \"N.A.\", SUPER_PATHWAY),\n",
    "                  SUB_PATHWAY=ifelse(SUB_PATHWAY==\"\", \"N.A.\", SUB_PATHWAY))\n",
    "\n",
    "head(node.tbl)\n",
    "tail(node.tbl)\n",
    "summary(node.tbl)\n",
    "print(str_c(\"Unique SUPER_PATHWAY: \", length(unique(node.tbl$SUPER_PATHWAY))))\n",
    "temp.tbl <- node.tbl %>%\n",
    "    dplyr::group_by(SUPER_PATHWAY) %>%\n",
    "    dplyr::summarize(Count=n()) %>%\n",
    "    dplyr::arrange(desc(Count))\n",
    "temp.tbl\n",
    "print(str_c(\"Unique SUB_PATHWAY: \", length(unique(node.tbl$SUB_PATHWAY))))\n",
    "temp.tbl <- node.tbl %>%\n",
    "    dplyr::group_by(SUPER_PATHWAY, SUB_PATHWAY) %>%\n",
    "    dplyr::summarize(Count=n()) %>%\n",
    "    dplyr::ungroup() %>%\n",
    "    dplyr::arrange(desc(Count))\n",
    "temp <- unique(temp.tbl$SUPER_PATHWAY)\n",
    "for (pathway in temp){\n",
    "    print(pathway)\n",
    "    temp.tbl %>% dplyr::filter(SUPER_PATHWAY==pathway) %>% print()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> • Use SUB_PATHWAY name if the category has more than 4 metabolites; otherwise, combine to \"Others (SUPER_PATHWAY)\".  \n",
    "> • \"Carbohydrate\", \"Nucleotide\", \"Cofactors and Vitamins\", \"Energy\", \"Peptide\", and \"Xenobiotics\" are combined to \"Others (Carbohydrate, Nucleotide, etc.)\".  \n",
    "> • Proteomics and Clinical labs are filled with label manually.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Combine small SUPER_PATHWAYs\n",
    "temp.c <- c(\"Carbohydrate\", \"Nucleotide\", \"Cofactors and Vitamins\", \"Energy\", \"Peptide\", \"Xenobiotics\")\n",
    "node.tbl <- node.tbl %>%\n",
    "    dplyr::mutate(SUPER_PATHWAY=ifelse((SUPER_PATHWAY %in% temp.c),\n",
    "                                       \"Others (Carbohydrate, Nucleotide, etc.)\", SUPER_PATHWAY))\n",
    "\n",
    "#Clean subcategory for the SUB_PATHWAY with <4 metabolites\n",
    "temp.tbl <- node.tbl %>%\n",
    "    dplyr::group_by(SUPER_PATHWAY, SUB_PATHWAY) %>%\n",
    "    dplyr::summarize(Count=n()) %>%\n",
    "    dplyr::ungroup() %>%\n",
    "    dplyr::arrange(desc(Count)) %>%\n",
    "    dplyr::filter(Count>=4)#Including proteomics and clinical labs as NA; and empty \"N.A.\"\n",
    "node.tbl <- node.tbl %>%\n",
    "    dplyr::mutate(Subcategory=ifelse((SUB_PATHWAY %in% unique(temp.tbl$SUB_PATHWAY)),\n",
    "                                     SUB_PATHWAY, str_c(\"Others (\", SUPER_PATHWAY, \")\"))) %>%\n",
    "    #dplyr::mutate(Subcategory=ifelse((SUPER_PATHWAY==\"N.A.\"), \"N.A.\", Subcategory)) %>%\n",
    "    dplyr::mutate(Subcategory=ifelse((SUPER_PATHWAY==\"Others (Carbohydrate, Nucleotide, etc.)\"),\n",
    "                                     \"Others (Carbohydrate, Nucleotide, etc.)\", Subcategory)) %>%\n",
    "    dplyr::mutate(Subcategory=ifelse((Category==\"Proteomics\"), \"Protein\", Subcategory)) %>%\n",
    "    dplyr::mutate(Subcategory=ifelse((Category==\"Clinical labs\"), \"Clinical Lab Test\", Subcategory))\n",
    "\n",
    "#Convert inconsistent capitalization of label\n",
    "node.tbl <- node.tbl %>%\n",
    "    dplyr::mutate(Subcategory=ifelse(Subcategory==\"Urea cycle; Arginine and Proline Metabolism\",\n",
    "                                     \"Urea Cycle; Arginine and Proline Metabolism\", Subcategory))\n",
    "\n",
    "print(str_c(\"nrows: \", nrow(node.tbl)))\n",
    "head(node.tbl)\n",
    "tail(node.tbl)\n",
    "print(str_c(\"Unique subcategory: \", length(unique(node.tbl$Subcategory))))\n",
    "temp.tbl <- node.tbl %>%\n",
    "    dplyr::group_by(Category, Subcategory) %>%\n",
    "    dplyr::summarize(Count=n()) %>%\n",
    "    dplyr::ungroup() %>%\n",
    "    dplyr::arrange(desc(Count))\n",
    "temp.tbl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-4. Prepare the sector table for visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The combined small SUPER_PATHWAYs are already updated in the SUPER_PATHWAY column.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sector.tbl <- node.tbl %>%\n",
    "    dplyr::select(Category, SUPER_PATHWAY, Subcategory) %>%\n",
    "    dplyr::distinct()\n",
    "\n",
    "#Numbering category order for visualization\n",
    "sector.tbl <- node.tbl %>%\n",
    "    dplyr::group_by(Category) %>%\n",
    "    dplyr::summarize(DummyCount=n()) %>%\n",
    "    dplyr::arrange(desc(DummyCount)) %>%\n",
    "    dplyr::mutate(CatNum=1:n()) %>%\n",
    "    dplyr::select(-DummyCount) %>%\n",
    "    dplyr::left_join(sector.tbl, ., by=\"Category\")\n",
    "\n",
    "#Numbering subcategory order for visualization (\"N.A.\" and \"Others (xxx)\" should be the last)\n",
    "sector.tbl <- node.tbl %>%\n",
    "    dplyr::group_by(SUPER_PATHWAY) %>%\n",
    "    dplyr::summarize(DummyCount=n()) %>%\n",
    "    dplyr::mutate(DummyCount=ifelse((SUPER_PATHWAY==\"N.A.\"|is.na(SUPER_PATHWAY)), NA, DummyCount)) %>%\n",
    "    dplyr::arrange(desc(DummyCount)) %>%\n",
    "    dplyr::mutate(DummyOrder=1:n()) %>%\n",
    "    dplyr::select(-DummyCount) %>%\n",
    "    dplyr::left_join(sector.tbl, ., by=\"SUPER_PATHWAY\")\n",
    "sector.tbl <- node.tbl %>%\n",
    "    dplyr::group_by(Subcategory) %>%\n",
    "    dplyr::summarize(Count=n()) %>%\n",
    "    dplyr::mutate(DummyCount=ifelse(str_detect(Subcategory, pattern=\"Others \"), NA, Count)) %>%\n",
    "    dplyr::left_join(sector.tbl, ., by=\"Subcategory\") %>%\n",
    "    dplyr::arrange(CatNum, DummyOrder, desc(DummyCount)) %>%\n",
    "    dplyr::group_by(Category) %>%\n",
    "    dplyr::mutate(SubcatNum=1:n()) %>%\n",
    "    dplyr::ungroup() %>%\n",
    "    dplyr::mutate(SectorNum=1:n()) %>%\n",
    "    dplyr::select(-DummyOrder, -DummyCount)\n",
    "sector.tbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "360/nrow(node.tbl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> When setting the gap degree b/w categories (dummy sectors in circos plot) to 5, x coordinate between subcategories (true sectors) should be less than 5 but more than 2.75.  \n",
    "> –> In this case, add 0.3 to Xcood as a gap between subcategories, while assuming the gap degree 3–4.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate the first/last x coordinate of subcategory on each category (dummy sector in circos plot)\n",
    "temp <- c()\n",
    "gap <- 0.3\n",
    "for (row_i in 1:nrow(sector.tbl)) {\n",
    "    subcat_i <- sector.tbl$SubcatNum[row_i]\n",
    "    if (subcat_i==1) {\n",
    "        temp <- c(temp, 1)\n",
    "    } else {\n",
    "        temp.tbl <- sector.tbl %>%\n",
    "            dplyr::filter(CatNum==sector.tbl$CatNum[row_i])\n",
    "        total_before <- sum(temp.tbl$Count[1:(subcat_i-1)])\n",
    "        temp <- c(temp, (total_before + gap*(subcat_i-1) + 1))\n",
    "    }\n",
    "}\n",
    "sector.tbl$SubcatXmin <- temp\n",
    "sector.tbl$SubcatXmax <- sector.tbl$SubcatXmin + sector.tbl$Count - 1\n",
    "\n",
    "#Calculate the first/last x corrdinate of category (dummy sector in circos plot)\n",
    "sector.tbl <- sector.tbl %>%\n",
    "    dplyr::group_by(Category) %>%\n",
    "    dplyr::summarize(CatXmin=min(SubcatXmin),\n",
    "                     CatXmax=max(SubcatXmax)) %>%\n",
    "    dplyr::left_join(sector.tbl, ., by=\"Category\")\n",
    "\n",
    "sector.tbl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-5. Calculate x coordinate of analyte"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate the x coordinate of node on each category (dummy sector in circos plot)\n",
    "node.tbl <- sector.tbl %>%\n",
    "    dplyr::select(Subcategory, SectorNum, SubcatXmin, SubcatXmax) %>%\n",
    "    dplyr::left_join(node.tbl, ., by=\"Subcategory\") %>%\n",
    "    dplyr::arrange(SectorNum, desc(nEdges), Analyte) %>%\n",
    "    dplyr::group_by(Subcategory) %>%\n",
    "    dplyr::mutate(DummyXcoord=1:n()) %>%\n",
    "    dplyr::ungroup() %>%\n",
    "    dplyr::mutate(Xcoord=SubcatXmin+DummyXcoord-1) %>%\n",
    "    dplyr::select(-DummyXcoord)\n",
    "head(node.tbl)\n",
    "print(\"Check margin\")\n",
    "node.tbl[16:26,]\n",
    "node.tbl[80:85,]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-6. Calculate label offset amount"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Skip this step because all labels are used this case.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-7. Circos plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check\n",
    "head(sector.tbl)\n",
    "head(node.tbl)\n",
    "head(pair.tbl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=10)#Default=7x7\n",
    "circos.clear()\n",
    "\n",
    "#Generate xlim table\n",
    "sector.tbl.cat <- sector.tbl %>%\n",
    "    dplyr::select(Category, CatNum, CatXmin, CatXmax) %>%\n",
    "    dplyr::distinct()\n",
    "\n",
    "#Initialize\n",
    "clockwise <- TRUE\n",
    "circos.par(start.degree=88, gap.degree=4, track.margin=c(0, 0),\n",
    "           cell.padding=c(0, 0, 0, 0), unit.circle.segments=nrow(node.tbl),\n",
    "           canvas.xlim=c(-2.3, 2.7), canvas.ylim=c(-2.35, 2.65),#Change based on the label length\n",
    "           clock.wise=clockwise, xaxis.clock.wise=clockwise)\n",
    "circos.initialize(sectors=sector.tbl.cat$CatNum, xlim=sector.tbl.cat[, c(\"CatXmin\", \"CatXmax\")],\n",
    "                  sector.width=sector.tbl.cat$CatXmax)\n",
    "\n",
    "#Prepare color palette with dummy sector order\n",
    "color_palette <- brewer.pal(nrow(sector.tbl.cat), \"Set1\")\n",
    "color_palette <- darken(color_palette, amount=0.05)\n",
    "\n",
    "#1st dummy track\n",
    "circos.track(ylim=c(0, 1), track.height=0.02, bg.col=NA, bg.border=NA)\n",
    "for (row_i in 1:nrow(node.tbl)) {\n",
    "    #Add node label\n",
    "    category <- node.tbl$Category[row_i]\n",
    "    sector <- sector.tbl.cat$CatNum[sector.tbl.cat$Category==category]\n",
    "    xcoord <- node.tbl$Xcoord[row_i]\n",
    "    label <- node.tbl$Analyte[row_i]\n",
    "    if (category==\"Metabolomics\") {\n",
    "        text_color <- color_palette[2]#blue\n",
    "    } else if (category==\"Proteomics\") {\n",
    "        text_color <- color_palette[1]#red\n",
    "    } else if (category==\"Clinical labs\") {\n",
    "        text_color <- color_palette[3]#green\n",
    "    }        \n",
    "    if (node.tbl$Highlight[row_i]==1) {\n",
    "        text_cex <- 0.75\n",
    "        text_style <- 2#bold font\n",
    "    } else if (node.tbl$Highlight[row_i]==0) {\n",
    "        text_cex <- 0.4\n",
    "        text_style <- 1\n",
    "    }\n",
    "    circos.text(x=xcoord, y=1, labels=label, sector.index=sector, track.index=1,\n",
    "                facing=\"clockwise\", niceFacing=TRUE, adj=c(0, 0.5),\n",
    "                cex=text_cex, col=text_color, font=text_style)\n",
    "    #Add axis ticks manually (circos.axis is not suitable in this case)\n",
    "    ##<- Ticks poisition can be adjusted, but the baseline is remained.\n",
    "    line_height <- 0.75\n",
    "    line_color <- \"black\"\n",
    "    line_width <- 1\n",
    "    circos.lines(x=rep(xcoord, 2), y=c(0, line_height),\n",
    "                 sector.index=sector, track.index=1, col=line_color, lwd=line_width)\n",
    "}\n",
    "\n",
    "#2nd dummy track\n",
    "circos.track(ylim=c(0, 1), track.height=0.1, bg.col=NA, bg.border=NA)\n",
    "for (row_i in 1:nrow(sector.tbl)) {\n",
    "    sector <- sector.tbl$CatNum[row_i]\n",
    "    xstart <- sector.tbl$SubcatXmin[row_i]\n",
    "    xend <- sector.tbl$SubcatXmax[row_i]\n",
    "    theta_start <- circlize(x=xstart, y=0, sector.index=sector, track.index=2)[1, 1]\n",
    "    theta_end <- circlize(x=xend, y=0, sector.index=sector, track.index=2)[1, 1]\n",
    "    #Add subcategory (true) sectors\n",
    "    if (sector.tbl$Category[row_i]==\"Metabolomics\") {\n",
    "        fill_color <- color_palette[2]#blue\n",
    "    } else if (sector.tbl$Category[row_i]==\"Proteomics\") {\n",
    "        fill_color <- color_palette[1]#red\n",
    "    } else if (sector.tbl$Category[row_i]==\"Clinical labs\") {\n",
    "        fill_color <- color_palette[3]#green\n",
    "    }\n",
    "    draw.sector(start.degree=theta_start, end.degree=theta_end, clock.wise=clockwise,\n",
    "                rou1=get.cell.meta.data(\"cell.top.radius\", track.index=2),\n",
    "                rou2=get.cell.meta.data(\"cell.bottom.radius\", track.index=2),\n",
    "                col=fill_color, border = \"black\", lwd=2)\n",
    "    #Add subcategory number\n",
    "    xcenter <- (xstart + xend) / 2\n",
    "    label <- str_c(\"#\", as.character(sector.tbl$SectorNum[row_i]))\n",
    "    circos.text(x=xcenter, y=0.5, labels=label, sector.index=sector, track.index=2,\n",
    "                facing=\"bending.inside\", niceFacing=TRUE, adj=c(0.5, 0.5),\n",
    "                cex=0.8, col=\"white\", font=2)#2 = bold font\n",
    "}\n",
    "\n",
    "#Add links\n",
    "for (row_i in 1:nrow(pair.tbl)) {\n",
    "    var1 <- pair.tbl$Variable1[row_i]\n",
    "    var2 <- pair.tbl$Variable2[row_i]\n",
    "    var1_category <- node.tbl$Category[node.tbl$Analyte==var1]\n",
    "    var2_category <- node.tbl$Category[node.tbl$Analyte==var2]\n",
    "    var1_sector <- sector.tbl.cat$CatNum[sector.tbl.cat$Category==var1_category]\n",
    "    var2_sector <- sector.tbl.cat$CatNum[sector.tbl.cat$Category==var2_category]\n",
    "    var1_xcoord <- node.tbl$Xcoord[node.tbl$Analyte==var1]\n",
    "    var2_xcoord <- node.tbl$Xcoord[node.tbl$Analyte==var2]\n",
    "    if (pair.tbl$GLM_Bcoef[row_i]>0) {\n",
    "        if (pair.tbl$Highlight[row_i]==1) {\n",
    "            line_color=rgb(1, 0, 0, alpha=0.6)\n",
    "            line_width=1\n",
    "        } else if (pair.tbl$Highlight[row_i]==0) {\n",
    "            line_color=rgb(1, 0, 0, alpha=0.2)\n",
    "            line_width=0.3\n",
    "        }\n",
    "    } else if (pair.tbl$GLM_Bcoef[row_i]<0) {\n",
    "        if (pair.tbl$Highlight[row_i]==1) {\n",
    "            line_color=rgb(0, 0, 1, alpha=0.6)\n",
    "            line_width=1\n",
    "        } else if (pair.tbl$Highlight[row_i]==0) {\n",
    "            line_color=rgb(0, 0, 1, alpha=0.2)\n",
    "            line_width=0.3\n",
    "        }\n",
    "    }\n",
    "    circos.link(sector.index1=var1_sector, point1=var1_xcoord,\n",
    "                sector.index2=var2_sector, point2=var2_xcoord,\n",
    "                h.ratio=0.6, w=4, col=line_color, lwd=line_width)\n",
    "}\n",
    "\n",
    "circos.clear()\n",
    "options(repr.plot.width=5, repr.plot.height=5)#Default=7x7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Because circlize package uses base graphycs, use graphycs device.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=10)#Default=7x7\n",
    "circos.clear()\n",
    "\n",
    "#To save (circlize uses base graphycs)\n",
    "fileDir = \"./ExportFigures/\"\n",
    "ipynbName <- \"220806_Multiomics-BMI-NatMed1stRevision_NetworkAnalysisVisualization_\"\n",
    "fileName <- \"longitudinal.pdf\"\n",
    "pdf(str_c(fileDir, ipynbName, fileName), 10, 10)\n",
    "\n",
    "#Generate xlim table\n",
    "sector.tbl.cat <- sector.tbl %>%\n",
    "    dplyr::select(Category, CatNum, CatXmin, CatXmax) %>%\n",
    "    dplyr::distinct()\n",
    "\n",
    "#Initialize\n",
    "clockwise <- TRUE\n",
    "circos.par(start.degree=88, gap.degree=4, track.margin=c(0, 0),\n",
    "           cell.padding=c(0, 0, 0, 0), unit.circle.segments=nrow(node.tbl),\n",
    "           canvas.xlim=c(-2.3, 2.7), canvas.ylim=c(-2.35, 2.65),#Change based on the label length\n",
    "           clock.wise=clockwise, xaxis.clock.wise=clockwise)\n",
    "circos.initialize(sectors=sector.tbl.cat$CatNum, xlim=sector.tbl.cat[, c(\"CatXmin\", \"CatXmax\")],\n",
    "                  sector.width=sector.tbl.cat$CatXmax)\n",
    "\n",
    "#Prepare color palette with dummy sector order\n",
    "color_palette <- brewer.pal(nrow(sector.tbl.cat), \"Set1\")\n",
    "color_palette <- darken(color_palette, amount=0.05)\n",
    "\n",
    "#1st dummy track\n",
    "circos.track(ylim=c(0, 1), track.height=0.02, bg.col=NA, bg.border=NA)\n",
    "for (row_i in 1:nrow(node.tbl)) {\n",
    "    #Add node label\n",
    "    category <- node.tbl$Category[row_i]\n",
    "    sector <- sector.tbl.cat$CatNum[sector.tbl.cat$Category==category]\n",
    "    xcoord <- node.tbl$Xcoord[row_i]\n",
    "    label <- node.tbl$Analyte[row_i]\n",
    "    if (category==\"Metabolomics\") {\n",
    "        text_color <- color_palette[2]#blue\n",
    "    } else if (category==\"Proteomics\") {\n",
    "        text_color <- color_palette[1]#red\n",
    "    } else if (category==\"Clinical labs\") {\n",
    "        text_color <- color_palette[3]#green\n",
    "    }        \n",
    "    if (node.tbl$Highlight[row_i]==1) {\n",
    "        text_cex <- 0.75\n",
    "        text_style <- 2#bold font\n",
    "    } else if (node.tbl$Highlight[row_i]==0) {\n",
    "        text_cex <- 0.4\n",
    "        text_style <- 1\n",
    "    }\n",
    "    circos.text(x=xcoord, y=1, labels=label, sector.index=sector, track.index=1,\n",
    "                facing=\"clockwise\", niceFacing=TRUE, adj=c(0, 0.5),\n",
    "                cex=text_cex, col=text_color, font=text_style)\n",
    "    #Add axis ticks manually (circos.axis is not suitable in this case)\n",
    "    ##<- Ticks poisition can be adjusted, but the baseline is remained.\n",
    "    line_height <- 0.75\n",
    "    line_color <- \"black\"\n",
    "    line_width <- 1\n",
    "    circos.lines(x=rep(xcoord, 2), y=c(0, line_height),\n",
    "                 sector.index=sector, track.index=1, col=line_color, lwd=line_width)\n",
    "}\n",
    "\n",
    "#2nd dummy track\n",
    "circos.track(ylim=c(0, 1), track.height=0.1, bg.col=NA, bg.border=NA)\n",
    "for (row_i in 1:nrow(sector.tbl)) {\n",
    "    sector <- sector.tbl$CatNum[row_i]\n",
    "    xstart <- sector.tbl$SubcatXmin[row_i]\n",
    "    xend <- sector.tbl$SubcatXmax[row_i]\n",
    "    theta_start <- circlize(x=xstart, y=0, sector.index=sector, track.index=2)[1, 1]\n",
    "    theta_end <- circlize(x=xend, y=0, sector.index=sector, track.index=2)[1, 1]\n",
    "    #Add subcategory (true) sectors\n",
    "    if (sector.tbl$Category[row_i]==\"Metabolomics\") {\n",
    "        fill_color <- color_palette[2]#blue\n",
    "    } else if (sector.tbl$Category[row_i]==\"Proteomics\") {\n",
    "        fill_color <- color_palette[1]#red\n",
    "    } else if (sector.tbl$Category[row_i]==\"Clinical labs\") {\n",
    "        fill_color <- color_palette[3]#green\n",
    "    }\n",
    "    draw.sector(start.degree=theta_start, end.degree=theta_end, clock.wise=clockwise,\n",
    "                rou1=get.cell.meta.data(\"cell.top.radius\", track.index=2),\n",
    "                rou2=get.cell.meta.data(\"cell.bottom.radius\", track.index=2),\n",
    "                col=fill_color, border = \"black\", lwd=2)\n",
    "    #Add subcategory number\n",
    "    xcenter <- (xstart + xend) / 2\n",
    "    label <- str_c(\"#\", as.character(sector.tbl$SectorNum[row_i]))\n",
    "    circos.text(x=xcenter, y=0.5, labels=label, sector.index=sector, track.index=2,\n",
    "                facing=\"bending.inside\", niceFacing=TRUE, adj=c(0.5, 0.5),\n",
    "                cex=0.8, col=\"white\", font=2)#2 = bold font\n",
    "}\n",
    "\n",
    "#Add links\n",
    "for (row_i in 1:nrow(pair.tbl)) {\n",
    "    var1 <- pair.tbl$Variable1[row_i]\n",
    "    var2 <- pair.tbl$Variable2[row_i]\n",
    "    var1_category <- node.tbl$Category[node.tbl$Analyte==var1]\n",
    "    var2_category <- node.tbl$Category[node.tbl$Analyte==var2]\n",
    "    var1_sector <- sector.tbl.cat$CatNum[sector.tbl.cat$Category==var1_category]\n",
    "    var2_sector <- sector.tbl.cat$CatNum[sector.tbl.cat$Category==var2_category]\n",
    "    var1_xcoord <- node.tbl$Xcoord[node.tbl$Analyte==var1]\n",
    "    var2_xcoord <- node.tbl$Xcoord[node.tbl$Analyte==var2]\n",
    "    if (pair.tbl$GLM_Bcoef[row_i]>0) {\n",
    "        if (pair.tbl$Highlight[row_i]==1) {\n",
    "            line_color=rgb(1, 0, 0, alpha=0.6)\n",
    "            line_width=1\n",
    "        } else if (pair.tbl$Highlight[row_i]==0) {\n",
    "            line_color=rgb(1, 0, 0, alpha=0.2)\n",
    "            line_width=0.3\n",
    "        }\n",
    "    } else if (pair.tbl$GLM_Bcoef[row_i]<0) {\n",
    "        if (pair.tbl$Highlight[row_i]==1) {\n",
    "            line_color=rgb(0, 0, 1, alpha=0.6)\n",
    "            line_width=1\n",
    "        } else if (pair.tbl$Highlight[row_i]==0) {\n",
    "            line_color=rgb(0, 0, 1, alpha=0.2)\n",
    "            line_width=0.3\n",
    "        }\n",
    "    }\n",
    "    circos.link(sector.index1=var1_sector, point1=var1_xcoord,\n",
    "                sector.index2=var2_sector, point2=var2_xcoord,\n",
    "                h.ratio=0.6, w=4, col=line_color, lwd=line_width)\n",
    "}\n",
    "\n",
    "#Close\n",
    "dev.off()\n",
    "\n",
    "circos.clear()\n",
    "options(repr.plot.width=5, repr.plot.height=5)#Default=7x7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> The exported figure is surely .pdf file.  \n",
    "> –> Furthermore, although different font was used on the output cell in Jupyter notebook, the Arial font was correctly used in the exported file!!  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — Session information —"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.1 (2021-08-10)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.3 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /opt/conda/envs/arivale-r/lib/libopenblasp-r0.3.18.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] arivale.data.interface_0.0.0.1 colorspace_2.0-3              \n",
       " [3] RColorBrewer_1.1-3             circlize_0.4.15               \n",
       " [5] forcats_0.5.1                  stringr_1.4.0                 \n",
       " [7] dplyr_1.0.9                    purrr_0.3.4                   \n",
       " [9] readr_2.1.2                    tidyr_1.2.0                   \n",
       "[11] tibble_3.1.7                   ggplot2_3.3.6                 \n",
       "[13] tidyverse_1.3.1               \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] lubridate_1.8.0     assertthat_0.2.1    digest_0.6.29      \n",
       " [4] utf8_1.2.2          IRdisplay_1.1       R6_2.5.1           \n",
       " [7] cellranger_1.1.0    repr_1.1.4          backports_1.4.1    \n",
       "[10] reprex_2.0.1        evaluate_0.15       httr_1.4.3         \n",
       "[13] pillar_1.7.0        GlobalOptions_0.1.2 rlang_1.0.2        \n",
       "[16] uuid_1.1-0          readxl_1.4.0        data.table_1.14.2  \n",
       "[19] rstudioapi_0.13     bit_4.0.4           munsell_0.5.0      \n",
       "[22] broom_0.8.0         compiler_4.1.1      modelr_0.1.8       \n",
       "[25] pkgconfig_2.0.3     base64enc_0.1-3     shape_1.4.6        \n",
       "[28] htmltools_0.5.2     tidyselect_1.1.2    fansi_1.0.3        \n",
       "[31] crayon_1.5.1        tzdb_0.3.0          dbplyr_2.2.0       \n",
       "[34] withr_2.5.0         grid_4.1.1          jsonlite_1.8.0     \n",
       "[37] gtable_0.3.0        lifecycle_1.0.1     DBI_1.1.2          \n",
       "[40] magrittr_2.0.3      scales_1.2.0        vroom_1.5.7        \n",
       "[43] cli_3.3.0           stringi_1.7.6       fs_1.5.2           \n",
       "[46] xml2_1.3.3          ellipsis_0.3.2      generics_0.1.2     \n",
       "[49] vctrs_0.4.1         IRkernel_1.3        tools_4.1.1        \n",
       "[52] bit64_4.0.5         glue_1.6.2          hms_1.1.1          \n",
       "[55] parallel_4.1.1      fastmap_1.1.0       rvest_1.0.2        \n",
       "[58] pbdZMQ_0.3-7        haven_2.5.0        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "arivale-r - R",
   "language": "R",
   "name": "conda-env-arivale-r-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
